File: web/checkpoint-controls.js
"use strict";
const checkpointDecisions=new Map(),checkpointDrafts=new Map(),checkpointPreviews=new Map(),checkpointDecisionPending=new Map(),checkpointDecisionDetails=new Map();
function checkpointDecisionDisclosure(key,label){
  const details=el('details'),id=workspaceId+':'+key;details.open=checkpointDecisionDetails.get(id)||false;
  details.append(el('summary',label));details.addEventListener('toggle',()=>checkpointDecisionDetails.set(id,details.open));return details;
}
function checkpointDecisionCurrent(entry){return entry&&entry.generation===workspaceGeneration;}
function checkpointDecisionStale(report){return report.workspaceRevision!==state.meta.revision||!(Date.now()/1000-report.inspectedAt>=0&&Date.now()/1000-report.inspectedAt<=60);}
function checkpointDecisionView(root){
  const report=checkpointDecisions.get(workspaceId),pending=checkpointDecisionPending.has(workspaceId),panel=el('section',null,'mission-status');
  panel.append(el('p',state.workspace.name+' · OWNER DECISION','eyebrow'),el('h2','Decide what may come next'),
    el('p','Review the saved checkpoint and exact next mission before a future run can be authorized. This decision does not accept a phase, start work, resume the brain or enable Play.'));
  const inspect=button(pending?'Inspecting saved decisions…':report?'Inspect decisions again':'Inspect checkpoint decisions',inspectCheckpointDecisions,'primary');inspect.disabled=pending||busy||!connected;panel.append(inspect);root.append(panel);
  if(!report){root.append(el('p','No decision inspection in this tab. Opening this page does not inspect proofs or change authority.','muted'));return;}
  root.append(el('p',`${checkpointDecisionStale(report)?'Earlier inspection · inspect again':'Saved local evidence · not live status'} · ${when(report.inspectedAt)}`,'muted'));
  if(report.status==='unavailable'){root.append(callout('Decision evidence unavailable','Retained review or withdrawal evidence is missing, changed or exceeds inspection limits. Ask the brain/operator to reconcile it. No fallback authority is offered.'));return;}
  if(!report.canReview)root.append(callout('Next-phase review unavailable',report.reviewBlocker));
  let entry=checkpointPreviews.get(workspaceId);
  if(entry&&!checkpointDecisionCurrent(entry)){checkpointPreviews.delete(workspaceId);entry=null;}
  if(entry){selected='checkpoint-confirm';checkpointDecisionConfirmation(root,entry);}
  else if(report.canReview)checkpointReviewEditor(root,report);
  const history=checkpointDecisionDisclosure('reviews',`Recorded owner reviews · ${report.reviews.length}`);history.append(
    el('p','Newest first. A new review does not cancel earlier reviews. Withdraw each unwanted review explicitly. Withdrawal blocks a future run grant; it does not stop a run already authorized. Use Pause for that.','muted'));
  for(const row of report.reviews){
    const item=el('section');item.append(el('h3',row.withdrawn?'Withdrawn review':'Recorded review · not running authority'),el('p',`Reviewed ${when(row.at)} · next-intent expiry ${when(row.expiresAt)}`),el('p','Review SHA-256: '+row.checkpointReviewHash,'mono mission-hash'),el('p','Report SHA-256: '+row.reportHash,'mono mission-hash'));
    if(row.withdrawn)item.append(el('p','Withdrawal recorded '+when(row.withdrawnAt)));
    else if(!entry)checkpointWithdrawalEditor(item,report,row);
    history.append(item);
  }
  if(!report.reviews.length)history.append(el('p','No owner checkpoint review has been recorded.'));
  root.append(history);
}
function checkpointReviewEditor(root,report){
  const c=report.candidate,key=workspaceId+':review';
  if(!checkpointDrafts.has(key))checkpointDrafts.set(key,{settings:'',minutes:''});
  const draft=checkpointDrafts.get(key),form=el('form',null,'mission-form');
  form.append(section('Review the next mission',`Saved report v${c.reportVersion} · generation ${c.generation} · ${when(c.reportAt)}`),el('p','Inspect the full saved report and next mission below before confirming. Counts are recorded claims, not phase acceptance.'));
  const spec=c.mission.spec;
  form.append(el('h3',spec.phase.title),el('p',spec.phase.objective),el('p','Required checkpoint: '+spec.phase.checkpoint),
    table(['Next mission limit','Reviewed value'],[['Approval mode',spec.authority.approvalMode],['Parallel tasks',num(spec.authority.maxParallelTasks)],['Total phase tasks',num(spec.authority.maxTasks)],['Phase token limit',num(spec.authority.tokenBudget)],['Included checkpoint reserve',num(spec.authority.checkpointReserveTokens)]]));
  const read=button('Inspect the bound report',()=>loadPhaseCheckpoints({reportHash:c.reportHash,artifactId:c.artifactId}));read.disabled=busy||!connected;form.append(read);
  const detail=checkpointDecisionDisclosure('mission:'+c.missionHash,'Next mission, scope and authority limits');detail.append(el('pre',JSON.stringify(c.mission,null,2)));form.append(detail);
  const label=el('label','Settings for the next intent'),select=el('select');select.required=true;
  const blank=el('option','Choose settings explicitly');blank.value='';select.append(blank);
  for(const [i,option] of c.settings.entries()){const node=el('option',option.label);node.value=String(i);select.append(node);}
  // A changed inspection must not reinterpret an old option index as a different policy.
  const selectedIndex=c.settings.findIndex(s=>JSON.stringify(s.value)===draft.settings);
  select.value=selectedIndex<0?'':String(selectedIndex);
  select.onchange=()=>{draft.settings=select.value===''?'':JSON.stringify(c.settings[Number(select.value)].value);selected='checkpoint-edit';};label.append(select);form.append(label);
  const duration=el('label','Next-intent validity (minutes)'),input=el('input');input.type='number';input.min=1;input.max=1440;input.step=1;input.required=true;input.value=draft.minutes;
  input.oninput=()=>{draft.minutes=input.value;selected='checkpoint-edit';};duration.append(input);form.append(duration,el('p','1–1,440 minutes from preview creation. This limits a future authorization; it is not a schedule or permission to run.','muted'));
  const submit=el('button','Preview checkpoint review');submit.type='submit';submit.disabled=busy||!connected||checkpointDecisionStale(report);form.append(submit);
  form.onsubmit=event=>{event.preventDefault();if(!form.reportValidity()||select.value==='')return;
    previewCheckpointDecision({operation:'review',expectedRevision:report.workspaceRevision,contextHash:report.contextHash,
      reportHash:c.reportHash,artifactId:c.artifactId,missionHash:c.missionHash,reviewReceiptHash:c.reviewReceiptHash,
      settingsPolicy:c.settings[Number(select.value)].value,expiresInSeconds:Number(input.value)*60});
  };root.append(form);
}
function checkpointWithdrawalEditor(root,report,row){
  const key=workspaceId+':'+row.checkpointReviewHash;
  if(!checkpointDrafts.has(key))checkpointDrafts.set(key,{reason:''});
  const draft=checkpointDrafts.get(key),form=el('form',null,'mission-form'),label=el('label','Reason to withdraw this review · no secrets'),input=el('textarea');
  input.rows=2;input.maxLength=2000;input.required=true;input.value=draft.reason;input.oninput=()=>{draft.reason=input.value;selected='checkpoint-edit';};label.append(input);form.append(label);
  const submit=el('button','Preview withdrawal');submit.type='submit';submit.disabled=busy||!connected||checkpointDecisionStale(report);form.append(submit);
  form.onsubmit=event=>{event.preventDefault();if(form.reportValidity())previewCheckpointDecision({operation:'withdraw',expectedRevision:report.workspaceRevision,contextHash:report.contextHash,checkpointReviewHash:row.checkpointReviewHash,reason:input.value});};root.append(form);
}
function checkpointDecisionConfirmation(root,entry){
  const d=entry.proposal.document,r=d.request,review=d.operation==='review',wrap=el('section',null,'mission-review retention-preview');
  wrap.id='checkpoint-decision-confirmation';wrap.tabIndex=-1;
  wrap.append(section(review?'Confirm checkpoint review':'Confirm withdrawal',`Preview expires ${when(d.expiresAt)}. Nothing has changed yet.`),
    table(['Exact scope','To be confirmed'],[['Workspace',d.workspaceId],['Report SHA-256',d.scope.reportHash],['Next mission SHA-256',d.scope.missionHash],...(review?[
      ['Settings',JSON.stringify(r.settingsPolicy)],['Next-intent expiry',when(r.expiresAt)]
    ]:[['Review SHA-256',r.checkpointReviewHash],['Reason',r.reason]])]),
    callout(review?'Review is not Play':'Already-authorized runs are unchanged',review?'This records your review of the exact checkpoint, next mission and settings. It does not authorize a run or clear admission, maintenance, native evidence or phase-acceptance gates.':'This withdraws only the selected review for future run grants. It does not stop or revoke a run already authorized; use safe Pause separately. Other reviews remain unchanged.'));
  const label=el('label'),check=el('input');check.type='checkbox';check.checked=false;label.append(check,el('span',review?'I reviewed this saved report, next mission and settings for the stated expiry.':'I want to withdraw this exact review, without stopping an already-authorized run.'));wrap.append(label);
  const submit=button(entry.uncertain?'Retry the same confirmation':review?'Record checkpoint review':'Withdraw this review',()=>{if(check.checked)confirmCheckpointDecision(entry);},'primary');submit.disabled=true;
  check.onchange=()=>{submit.disabled=busy||!connected||!check.checked;};
  wrap.append(submit,button('Back to decisions',()=>{if(busy)return;checkpointPreviews.delete(workspaceId);selected=null;render();}));
  if(entry.uncertain)wrap.append(el('p','The response was not confirmed. Retry uses the identical signed request; inspect saved decisions before preparing a different one.','muted'));
  const detail=el('details');detail.append(el('summary','Exact signed decision and mission'),el('pre',JSON.stringify(d,null,2)));wrap.append(detail);root.append(wrap);
}
async function inspectCheckpointDecisions(){
  const wid=workspaceId,generation=workspaceGeneration;
  if(!wid||busy||checkpointDecisionPending.has(wid)||!connected)return;
  const token={};checkpointDecisionPending.set(wid,token);checkpointPreviews.delete(wid);selected=null;render();
  try{const report=await api('/api/checkpoint-decisions');if(workspaceId===wid&&workspaceGeneration===generation&&report.workspaceId===wid)checkpointDecisions.set(wid,report);}
  catch(error){if(workspaceId===wid&&workspaceGeneration===generation&&!error.workspaceChanged){checkpointDecisions.delete(wid);showNotice('Decision inspection failed. No authority changed. '+error.message,true);}}
  finally{if(checkpointDecisionPending.get(wid)===token)checkpointDecisionPending.delete(wid);if(workspaceId===wid&&workspaceGeneration===generation&&view==='phaseCheckpoints')render();}
}
async function checkpointDecisionPost(path,payload,done){
  if(busy||!connected)return false;
  const wid=workspaceId,generation=workspaceGeneration;busy=true;updateWorkspaceSelector();if(view==='phaseCheckpoints')render();
  try{const result=await api(path,{method:'POST',headers:{'Content-Type':'application/json','X-CSRF-Token':csrf},body:JSON.stringify(payload)});
    if(workspaceId===wid&&workspaceGeneration===generation){done(result);return true;}
  }catch(error){if(!error.workspaceChanged&&workspaceId===wid&&workspaceGeneration===generation){
    if(path.endsWith('/confirm')&&checkpointPreviews.has(wid))checkpointPreviews.get(wid).uncertain=true;
    showNotice(error.message+' No automatic retry was sent. Your inputs are retained.',true);
  }}finally{busy=false;updateWorkspaceSelector();if(workspaceId===wid&&workspaceGeneration===generation&&view==='phaseCheckpoints')render();}
  return false;
}
async function previewCheckpointDecision(payload){
  const generation=workspaceGeneration;
  const saved=await checkpointDecisionPost('/api/checkpoint-decisions/preview',payload,proposal=>{
    if(proposal.document.workspaceId!==workspaceId)throw new Error('Preview belongs to another workspace');
    checkpointPreviews.set(workspaceId,{proposal,generation,uncertain:false});selected='checkpoint-confirm';
  });
  if(saved&&generation===workspaceGeneration&&view==='phaseCheckpoints'&&typeof document!=='undefined')document.getElementById('checkpoint-decision-confirmation')?.focus();
  return saved;
}
async function confirmCheckpointDecision(entry){
  if(!checkpointDecisionCurrent(entry)||entry!==checkpointPreviews.get(workspaceId))return false;
  const wid=workspaceId,generation=workspaceGeneration,review=entry.proposal.document.operation==='review';
  const saved=await checkpointDecisionPost('/api/checkpoint-decisions/confirm',{proposal:entry.proposal,confirmed:true},result=>{
    checkpointPreviews.delete(wid);checkpointDecisions.delete(wid);selected=null;
    showNotice(result.replayed?'Original decision receipt recovered; no authority was reapplied.':review?'Checkpoint review recorded. No run was authorized or started.':'Review withdrawn for future run grants. Already-authorized runs are unchanged.');
  });
  if(saved&&workspaceId===wid&&workspaceGeneration===generation){await refresh();await inspectCheckpointDecisions();}
  return saved;
}

File: orchestrator/checkpoint_controls.py
"""Authenticated owner adapter; no run grant, native call, notification or scheduler."""
import contextlib
import copy
import hmac
import sqlite3
import time
import uuid

from . import missions, model_policy, phase_checkpoints as phases, run_authority as runs, task_contracts
from .admission import exact, integer, sha
from .core import Refusal, canonical, digest, require
from .retention_controls import RetentionControls, identity, TTL

INVALID = (Refusal, KeyError, TypeError, ValueError, RecursionError)
REVIEW_INPUT = {"reportHash", "artifactId", "missionHash", "reviewReceiptHash", "settingsPolicy", "expiresInSeconds"}
BOUNDARY = "Owner review only. No run is authorized or resumed; admission, maintenance and native evidence gates remain."


def inspect_in(ledger, db):
    meta = ledger.get(db, "meta", 1)
    out = {"workspaceId": missions.workspace(ledger), "workspaceRevision": meta["revision"], "inspectedAt": time.time(),
           "status": "unavailable", "candidate": None, "reviews": [], "canReview": False,
           "reviewBlocker": "Checkpoint review history is unavailable; ask the operator to reconcile it.",
           "executionAuthorized": False, "nativeCallMade": False, "boundary": BOUNDARY}
    try:
        count, size = db.execute("SELECT count(*),coalesce(sum(length(CAST(data AS BLOB))),0) FROM snapshots WHERE kind=?", (phases.REVIEW,)).fetchone()
        require(count <= 128 and size <= 2_000_000, "Checkpoint review history needs bounded migration")
        withdrawn = phases.withdrawals_in(ledger, db)
        for row in db.execute("SELECT id FROM snapshots WHERE kind=? ORDER BY rowid DESC", (phases.REVIEW,)):
            record = phases.owner_review_in(ledger, db, row[0]); req = record["request"]
            out["reviews"].append({"checkpointReviewHash": row[0], "at": record["at"], "withdrawn": row[0] in withdrawn,
                "withdrawnAt": phases.document(db, withdrawn[row[0]], phases.WITHDRAWAL)["at"] if row[0] in withdrawn else None,
                **{k: req[k] for k in ("reportHash", "artifactId", "missionHash", "reviewReceiptHash", "settingsPolicy", "expiresAt")}})
        out["status"] = "recorded"
        try:
            pointer = meta.get("phaseCheckpointReport")
            require(pointer is not None, "No phase report is saved. Ask the designated brain to prepare one after a safe checkpoint.")
            exact(pointer, {"reportHash", "artifactId"})
            source = runs.mission_source(ledger, db); task_contracts.reviewed_phase(source)
            m = source["mission"]
            req = {**pointer, "missionHash": m["documentHash"], "reviewReceiptHash": m["receiptHash"],
                   "settingsPolicy": "native_defaults", "expiresAt": time.time()+3600, "confirmed": True}
            phases.validate_review_in(ledger, db, req)
            settings = [{"value": "native_defaults", "label": "Native defaults · no model override", "policy": None}]
            if meta.get("modelPolicyHash"):
                try:
                    value = {"mode": "adaptive", "policyHash": meta["modelPolicyHash"]}
                    policy = model_policy.policy_in(ledger, db, value)
                    settings.append({"value": value, "label": "Current owner-reviewed adaptive policy", "policy": policy})
                except INVALID: pass  # Invalid optional settings are never silently selected.
            report = phases.read_in(ledger, db, pointer["reportHash"], pointer["artifactId"])["report"]
            info, _ = phases.artifact_bytes(db, pointer["artifactId"])
            out["candidate"] = {**pointer, "reportVersion": info["version"], "generation": report["generation"],
                "reportAt": report["retainedAt"], "checkpointAt": report["checkpointAt"], "summary": report["summary"],
                "missionHash": m["documentHash"], "reviewReceiptHash": m["receiptHash"], "mission": m["document"], "settings": settings}
            require(count < 128, "Review history limit reached; explicit migration required before another review")
            out.update(canReview=True, reviewBlocker=None)
        except INVALID as error:
            out["reviewBlocker"] = str(error) if isinstance(error, Refusal) else "Saved checkpoint or next mission evidence is invalid. Ask the brain/operator to reconcile it."
    except INVALID:
        out.update(status="unavailable", candidate=None, reviews=[], canReview=False)
    out["contextHash"] = digest({k: out[k] for k in ("workspaceId", "workspaceRevision", "status", "candidate", "reviews", "canReview")})
    return out


def inspect(ledger):
    with contextlib.closing(sqlite3.connect(ledger.db.as_uri()+"?mode=ro", uri=True)) as db:
        db.row_factory = sqlite3.Row; db.execute("BEGIN")
        return inspect_in(ledger, db)


def summary(report, revision):
    if not report: return {"status": "not_inspected", "executionAuthorized": False}
    return {"status": report["status"], "workspaceRevision": report["workspaceRevision"], "inspectedAt": report["inspectedAt"],
            "historical": True, "workspaceChanged": revision != report["workspaceRevision"],
            "expired": not 0 <= time.time()-report["inspectedAt"] <= 60, "executionAuthorized": False,
            "recordedReviews": len(report["reviews"]), "recordedWithdrawals": sum(r["withdrawn"] for r in report["reviews"])}


class CheckpointControls(RetentionControls):
    """Reuse ephemeral signing only; each feature/workspace has its own random key."""
    def preview(self, ledger, body, session):
        body = copy.deepcopy(body)
        require(isinstance(body, dict) and body.get("operation") in ("review", "withdraw"), "Checkpoint review or withdrawal required")
        op = body["operation"]
        exact(body, {"operation", "expectedRevision", "contextHash", *(REVIEW_INPUT if op == "review" else {"checkpointReviewHash", "reason"})})
        integer(body["expectedRevision"]); sha(body["contextHash"])
        report = inspect(ledger)
        require(report["status"] == "recorded" and body["expectedRevision"] == report["workspaceRevision"] and
                body["contextHash"] == report["contextHash"], "Workspace changed; inspect checkpoint decisions again")
        now = time.time()
        request = {"id": str(uuid.uuid4()), "expectedRevision": body["expectedRevision"], "confirmed": True}
        if op == "review":
            require(report["canReview"], report["reviewBlocker"])
            c = report["candidate"]
            require(all(body[k] == c[k] for k in ("reportHash", "artifactId", "missionHash", "reviewReceiptHash")), "Exact inspected report and next mission required")
            require(any(body["settingsPolicy"] == s["value"] for s in c["settings"]), "Select an available settings policy explicitly")
            integer(body["expiresInSeconds"], 60, 86400)
            request.update({k: body[k] for k in REVIEW_INPUT if k != "expiresInSeconds"}, expiresAt=now+body["expiresInSeconds"])
            scope = c
        else:
            sha(body["checkpointReviewHash"]); missions.text(body["reason"], "Withdrawal reason")
            scope = next((r for r in report["reviews"] if r["checkpointReviewHash"] == body["checkpointReviewHash"]), None)
            require(scope and not scope["withdrawn"], "Exact non-withdrawn review required")
            request.update(checkpointReviewHash=body["checkpointReviewHash"], reason=body["reason"])
        doc = {"kind": "checkpoint_owner_preview", "operation": op, "request": request, "scope": scope,
               "workspaceId": report["workspaceId"], "ledgerIdentity": identity(ledger), "session": digest(session),
               "contextHash": report["contextHash"], "createdAt": now, "expiresAt": now+TTL, "boundary": BOUNDARY}
        require(len(canonical(doc).encode()) <= 28000, "Checkpoint preview exceeds its bound; narrow the reviewed mission")
        return {"document": doc, "signature": self.sign(doc)}

    def confirm(self, ledger, body, session):
        body = copy.deepcopy(body); exact(body, {"proposal", "confirmed"})
        require(body["confirmed"] is True, "Explicit owner confirmation required")
        exact(body["proposal"], {"document", "signature"})
        doc = body["proposal"]["document"]; signature = body["proposal"]["signature"]; sha(signature)
        require(hmac.compare_digest(signature, self.sign(doc)), "Preview changed or dashboard restarted; inspect saved decisions before preparing again")
        require(doc["kind"] == "checkpoint_owner_preview" and doc["session"] == digest(session) and
                doc["workspaceId"] == missions.workspace(ledger) and doc["ledgerIdentity"] == identity(ledger), "Preview belongs to another session or workspace")
        op, req = doc["operation"], doc["request"]
        fields = phases.REVIEW_FIELDS if op == "review" else {"checkpointReviewHash", "reason", "confirmed"}
        with ledger.tx() as db:
            _, _, _, prior = runs.request_in(ledger, db, req, "dashboard_owner", "phase_"+op, fields)
            if not prior:
                require(0 <= time.time()-doc["createdAt"] <= TTL and time.time() <= doc["expiresAt"], "Preview expired; inspect and preview again")
                require(inspect_in(ledger, db)["contextHash"] == doc["contextHash"], "Workspace changed; inspect checkpoint decisions again")
            method = phases.review_in if op == "review" else phases.withdraw_in
            receipt = method(ledger, db, req, actor="dashboard_owner")
            return {"receipt": receipt, "replayed": prior is not None, "executionAuthorized": False, "nativeCallMade": False}
